Add ios-apps extension

chrismessina opened the pull request:
## Description

<!-- A summary of your change. If you add a new extension or command, explain what it does. -->

## Screencast

<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

## Checklist

- [ ] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [ ] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [ ] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [ ] I checked that files in the `assets` folder are used by the extension itself
- [ ] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented:
Hi 👋

It seems a bit similar to https://www.raycast.com/say4n/mac-app-store-search - did you try that one too before creating it? They might use different APIs though

pernielsentikaer commented:
I guess this is another segment since the other one focus on Apps for macOS

chrismessina commented:
@pernielsentikaer yes — I did review it and considered modeling this after it, but after digging in, realized that they use different APIs and methods (mine uses the `ipatool` to facilitate downloading .ipa files). 

My extension is specifically for iOS Apps and doesn't interact with the Mac App Store APIs.

andreaselia commented:
@greptileai can you check this?

chrismessina commented:
I've addressed all of @greptileai's comments and added a function to download hi-res screenshots.

chrismessina commented on extensions/ios-apps/package.json:
```diff
@@ -0,0 +1,193 @@
+    "@raycast/api": "^1.93.1",
+    "@raycast/utils": "^1.17.0",
+    "@types/stream-json": "^1.7.8",
+    "axios": "^1.6.0",
+    "lodash": "^4.17.21",
+    "node-fetch": "^3.3.2",
```
Removed axios.

chrismessina commented on extensions/ios-apps/src/utils/common.ts:
```diff
@@ -0,0 +1,138 @@
+      // If we can't parse as JSON, check if it contains "Not authenticated"
+      if (stdout.includes("Not authenticated")) {
+        // If not authenticated, login
+        await execAsync(
+          `${IPATOOL_PATH} auth login -e "${preferences.appleId}" -p "${preferences.password}" --format json --non-interactive`,
+        );
```
I have extracted the Apple ID login logic into a new helper function `loginToAppleId()` in `common.ts` and replaced all duplicate execAsync login calls with calls to this function. This eliminates code repetition and improves maintainability, while preserving strict typing and clarity.

chrismessina commented on extensions/ios-apps/src/components/app-actions.tsx:
```diff
@@ -0,0 +1,56 @@
+  const appStoreUrl = app.trackViewUrl || (app.id ? `https://apps.apple.com/app/id${app.id}` : undefined);
+
+  // Default download handler if none provided
+  const handleDownload = async () => {
+    try {
+      if (onDownload) {
+        return await onDownload(app);
+      }
+
+      // Fall back to direct download if no handler provided
+      return await downloadIPA(app.bundleId, app.name, app.version, app.price);
+    } catch (error) {
+      console.error("Error downloading app:", error);
+      throw error;
+    }
+  };
```
Done!

chrismessina commented on extensions/ios-apps/src/hooks/use-app-screenshots.ts:
```diff
@@ -0,0 +1,69 @@
+        error: error instanceof Error ? error : new Error(String(error)),
+        downloadPath: null,
+      });
+      await showToast(Toast.Style.Failure, "Failed to download screenshots", String(error));
```
Done.

chrismessina commented on extensions/ios-apps/src/utils/itunes-api.ts:
```diff
@@ -0,0 +1,374 @@
+    console.log(`[iTunes API] Fetching app details for bundleId: ${bundleId}`);
+    const url = `https://itunes.apple.com/lookup?bundleId=${encodeURIComponent(bundleId)}&country=us&entity=software`;
+
+    const response = await fetch(url);
+    if (!response.ok) {
+      console.error(`[iTunes API] Request failed with status: ${response.status}`);
+      return null;
+    }
```
Done.

chrismessina commented on extensions/ios-apps/src/utils/common.ts:
```diff
@@ -0,0 +1,138 @@
+export function renderStarRating(rating: number | undefined): string {
+  if (rating === undefined) return "No Rating";
+
+  const fullStars = Math.floor(rating);
+  const halfStar = rating % 1 >= 0.5;
+  const emptyStars = 5 - fullStars - (halfStar ? 1 : 0);
```
The renderStarRating function now clamps the rating between 0 and 5, ensuring correct star rendering for any input value. This prevents display errors for out-of-bounds ratings and aligns with code quality and correctness best practices.

andreaselia commented:
Thanks @chrismessina, I'll be taking a look at this in the coming days.

The linter is currently failing. Once sorted, please mark the PR as ready to review.

andreaselia commented:
@chrismessina do you think it's worth including that preference as part of the initial release?

chrismessina commented:
@andreaselia this should be ready for review. I've made downloads optional behind an experimental setting!

andreaselia commented:
@chrismessina did you sort any branching issues out? Is this one good for a review with all of your changes in now?

chrismessina commented:
![2025-05-29 (15 38 55) Raycast](https://github.com/user-attachments/assets/39bebe88-0fc8-4461-b25c-4a66ae774703)

Yes, should be ready for a review.

The ipatool is still broken, so if you don't enable Experimental Downloads, you should still be able to search for iOS Apps and download screenshots.

andreaselia commented:
@chrismessina Fancy taking a look at Greptile's feedback and seeing if any of it applies to your PR? You can always just resolve the comments without action if you feel it doesn't make sense to apply certain changes.

chrismessina commented on raycast-extensions/ios-apps/README.md:
```diff
@@ -0,0 +1,141 @@
+
+- [ipatool](https://github.com/majd/ipatool) by Majd Alfhaily
+- [iTunes Search API](https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/iTuneSearchAPI/index.html) by Apple
+- [Windsurf](https://codeium.com/refer?referral_code=650bebd9a5) by Codeium
```
Ok fine.

chrismessina commented on raycast-extensions/ios-apps/README.md:
```diff
@@ -0,0 +1,141 @@
+
+- **Search**: Quickly search for iOS apps by name, developer, or bundle ID
+- **Rich App Details**: View comprehensive app information including ratings, screenshots, and metadata
+- **Download**: Download IPA files directly to your computer
```
This is resolved.

chrismessina commented:
@andreaselia When I addressed all of @greptile-bot's comments, I ended up breaking a bunch of things — particularly because it misunderstood how the ipatool was handing authentication on the command line. 

I fixed my regressions and also developed a much more robust way to get app screenshots.

I also added a toggle for each platform's screenshots, as well as the ability to toggle verbose logging.

Please take a look!

chrismessina commented on extensions/ios-apps/src/utils/auth.ts:
```diff
@@ -0,0 +1,59 @@
+ */
+async function loginToAppleId(): Promise<void> {
+  await execAsync(
+    `${IPATOOL_PATH} auth login -e "${preferences.appleId}" -p "${preferences.password}" --format json --non-interactive`,
```
No. This is what screwed me over last time. If Raycast Preferences are insecure, they should fix that. I'm using standard practices. 

https://developers.raycast.com/api-reference/preferences

chrismessina commented on raycast-extensions/ios-apps/src/hooks/use-clipboard.ts:
```diff
@@ -0,0 +1,22 @@
+   * @param toastTitle Title for the toast notification
+   * @param toastMessage Optional message for the toast notification
+   */
+  const copyToClipboard = async (text: string, toastTitle: string, toastMessage = "Copied to clipboard") => {
+    await Clipboard.copy(text);
+    showToast(Toast.Style.Success, toastTitle, toastMessage);
+  };
```
The existing implementation:

Has a try-catch block to handle any errors
Uses showFailureToast from @raycast/utils for error reporting
Passes the error object directly to showFailureToast with a descriptive title
The code is already following the best practice you're suggesting. Is there something specific about the error handling you'd like to modify or enhance?

chrismessina commented on extensions/ios-apps/src/search.tsx:
```diff
@@ -0,0 +1,116 @@
+        <List.EmptyView
+          title="Type Query to Search"
+          description="Search for apps by name, developer, or bundle Id."
+          icon="no-view.png"
```
This isn't a favicon.

chrismessina commented on raycast-extensions/ios-apps/package.json:
```diff
@@ -0,0 +1,115 @@
+      "mode": "view"
+    }
+  ],
+  "tools": [
+    {
+      "name": "download-app",
+      "title": "Download iOS App",
+      "description": "Download an iOS app"
+    },
+    {
+      "name": "get-app-details",
+      "title": "Get iOS App Details",
+      "description": "Get detailed information about an iOS app by name or search term"
+    },
+    {
+      "name": "search-apps",
+      "title": "Search iOS Apps",
+      "description": "Search for iOS apps on the App Store by name or keyword"
+    }
```
Nope. AI Evals are in ai.yaml. Please check yo'self.

chrismessina commented on raycast-extensions/ios-apps/package.json:
```diff
@@ -0,0 +1,115 @@
+      "description": "Search for iOS apps on the App Store by name or keyword"
+    }
+  ],
+  "preferences": [
+    {
+      "name": "appleId",
+      "title": "Apple ID Email",
+      "description": "Your Apple ID email for authentication",
+      "type": "textfield",
+      "required": false
+    },
+    {
+      "name": "password",
+      "title": "Apple ID Password",
+      "description": "Your Apple ID password for authentication",
+      "type": "password",
+      "required": false
+    },
```
I'm using Raycast preference store.

chrismessina commented on raycast-extensions/ios-apps/src/components/app-actions.tsx:
```diff
@@ -0,0 +1,89 @@
+  const preferences = getPreferenceValues<ExtensionPreferences>();
+  const isExperimentalDownloadsEnabled = preferences.enableAppDownloads || false;
+
+  // Create a fallback App Store URL if trackViewUrl is not available
+  const appStoreUrl = app.trackViewUrl || (app.id ? `https://apps.apple.com/app/id${app.id}` : undefined);
```
Fixed upstream.

chrismessina commented on extensions/ios-apps/src/tools/get-app-details.ts:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        // Prefer shorter names when scores are similar (likely more precise matches)
+        score -= name.length * 0.1;
```
Summary of Changes
Fixed score penalty logic in app-search.ts:

Replaced score -= name.length * 0.1 with a capped penalty system
Added lengthPenalty = Math.min(name.length * 0.1, 10) to cap the maximum penalty at 10 points
This ensures that even apps with very long names (100+ characters) won't have their scores go negative
Maintains the preference for shorter names while preventing scoring issues
Benefits of this fix:

Prevents negative scores: Even apps with 200+ character names will only lose a maximum of 10 points
Maintains ranking logic: Shorter names are still preferred, but within reasonable bounds
Preserves exact matches: A perfect match (100 points) will always score at least 90 points, regardless of name length
Consistent behavior: All scoring remains positive, making the ranking more predictable
This change ensures that the scoring algorithm behaves correctly for all app names, regardless of their length, while still maintaining the intended preference for shorter, more precise matches.

chrismessina commented on raycast-extensions/ios-apps/src/utils/paths.ts:
```diff
@@ -0,0 +1,27 @@
+
+// Get the downloads directory path from preferences or default to ~/Downloads
+export function getDownloadsDirectory(): string {
+  if (preferences.downloadPath) {
+    // Replace ~ with the actual home directory if present
+    if (preferences.downloadPath.startsWith("~")) {
+      return preferences.downloadPath.replace("~", homedir());
+    }
+    return preferences.downloadPath;
+  }
```
I've created a cleaner, more streamlined version that combines the best aspects of both approaches:

What Makes This Version Better:
Simpler Logic Flow: Uses ternary operators and combines operations where logical
Single Try-Catch Block: Handles both directory creation and writability validation in one place
Consistent Variable Naming: Uses downloadPath throughout for clarity
Maintains All Security: Still includes directory traversal protection
Robust Functionality: Creates missing directories and validates writability
Cleaner Error Handling: Single, comprehensive error message
Key Improvements Over Your Suggestion:
✅ Keeps Security Check: Directory traversal protection maintained
✅ Auto-Creates Directories: Won't fail if directory doesn't exist
✅ Consistent Imports: Uses existing ES6 import style
✅ Single Responsibility: One try-catch handles both creation and validation
✅ Better Error Messages: More descriptive error information

chrismessina commented:
LFG!!! Check it out @andreaselia!

chrismessina commented:
Yikes. You're seeing some strange stuff.

> Does it make sense in such a small extension like this to have a test environment? I'm thinking especially since it seems to mock rather than actually being able to ensure the request is correct, if I'm not mistaken? 🤔

Because of the way I'm gathering data for downloading the screenshots, Claude built out tests in case the scraping method breaks. If you think it's unnecessary, I can remove it. I left it in to facilitate maintenance down the road in case Apple updates the App Store webpages.
 
> There are duplicate images where re-using the extension icon would work.

Can you provide an example? I'm not sure what you're referring to.

 
> There's no way to handle two-factor authentication so login fails (see the Hide My Email extension for how they handle 2FA with an extra screen to enter the code).

Hmm. Ok. Not sure if that will apply here since we're getting an Apple auth token for use of the ipatool CLI. I'll look into it.

> 
> ![image](https://private-user-images.githubusercontent.com/5033092/464072942-efaded15-5636-408d-a0d1-325ee220155a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIwODgzNTAsIm5iZiI6MTc1MjA4ODA1MCwicGF0aCI6Ii81MDMzMDkyLzQ2NDA3Mjk0Mi1lZmFkZWQxNS01NjM2LTQwOGQtYTBkMS0zMjVlZTIyMDE1NWEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDcwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA3MDlUMTkwNzMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDI4YWRlYmQ5YzBlYWE1NmJjMDRhMjJiYzBiMzNkN2M3OTI3NWU0YTliMWM5MTYxYjIxNWY0NTRhZDRlMWY5YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T1sOby8tA45vIShbHH77RGhBKYakexQ_onB-SmcDeSo)

Did you install the ipatool and configure it in your settings?

> Search is very slow and buggy, sometimes not loading the results for the search term, and the state for loading seems a little off with the no results screen.

Can you provide a screen recording? 

When a search is first performed, it quickly returns basic results from `ipatool`, then progressively enriches each result with detailed data from the iTunes API. You should see a progress indicator during the enrichment phase, and the interface updates as more detailed information becomes available."


> Images don't work in the list view or when showing details.

I wonder if there's some geo restriction on what you're seeing because the images do load for me. 

![2025-07-09 (12 18 44) Arc](https://github.com/user-attachments/assets/8f6b3170-e72b-45c8-b9b8-041f3a3822ee)

chrismessina commented:
@greptile @greptile-apps could you take another look?

chrismessina commented on extensions/ios-apps/README.md:
```diff
@@ -0,0 +1,241 @@
+
+- [ipatool](https://github.com/majd/ipatool) by Majd Alfhaily
+- [iTunes Search API](https://developer.apple.com/library/archive/documentation/AudioVideo/Conceptual/iTuneSearchAPI/index.html) by Apple
+- [Windsurf](https://windsurf.com/)
```
That's not a referral URL. That's Windsurf's new homepage.

chrismessina commented on raycast-extensions/ios-apps/src/tools/search-apps.ts:
```diff
@@ -0,0 +1,82 @@
+        // Convert from ipatool format to AppDetails format
+        const appDetails = {
+          id: String(app.id),
+          bundleId: app.bundleID,
```
I know. We handle both bundleId and bundleID formats because of ipatool.

chrismessina commented on raycast-extensions/ios-apps/src/tools/search-apps.ts:
```diff
@@ -0,0 +1,82 @@
+
+    return { apps: formattedApps };
+  } catch (error) {
+    console.error(`[search-apps tool] Error: ${error}`);
+    throw new Error(`Failed to search for apps: ${error}`);
```
I fixed a similar problem elsewhere but can't find this code.

andreaselia commented on raycast-extensions/ios-apps/ai.yaml:
```diff
@@ -0,0 +1,161 @@
+        -  **Version:** 1.0.6
+        -  **Developer:** Raycast Technologies Inc
+        -  **Rating:** 4.38/5
+        -  **Description:** Your all-in-one productivity toolkit with AI, notes, and more — now on mobile as the perfect companion to the macOS app.
+        -  **Link:** [View on App Store](https://apps.apple.com/us/app/raycast/id6446901002)
```
Greptile imagining things again 😂

chrismessina commented:
@greptile could you take another look?

chrismessina commented:
@greptile could you take another look?

chrismessina commented:
@greptile could you take another look?

chrismessina commented on extensions/ios-apps/CONFIG.md:
```diff
@@ -0,0 +1,136 @@
+
+## Example Usage
+
+See `src/utils/config-example.ts` for comprehensive usage examples including:
+
+- Download functions with timeout handling
+- Integrity verification based on config
+- Cleanup logic based on preferences
+- React hook-style usage patterns
```
Restored!

chrismessina commented:
@greptile could you take another look?

pernielsentikaer commented:
I was only able to log in through the terminal. I didn't see any 2FA codes in the extension; is that intended?

Another issue I encountered that might be a bit of a showstopper for the download is that if you have the setting below enabled, the extension unloads as soon as Raycast closes (because it asks for the keychain).

<img width="2320" height="1620" alt="CleanShot 2025-09-02 at 16 02 44@2x" src="https://github.com/user-attachments/assets/1bca089f-03aa-4f7b-8f51-29f681da8e9c" />

I'm not sure if we can bypass that easily?

chrismessina commented:
> I was only able to log in through the terminal. I didn't see any 2FA codes in the extension; is that intended?

Can you describe your flow? Similarly, I haven't been able to cause the 2FA flow to trigger, even when I sign out completely and remove the Keychain credentials. 

This was a blocker that @andreaselia asked me to address, and so I followed [the approach in the Hide My Email extension](https://github.com/raycast/extensions/blob/fc54e0d0e82a3aaec0983dd4be128ce935acdbf0/extensions/hidemyemail/src/api/connect.ts), except that that uses iCloud APIs, whereas my extension does not. 

Ultimately I made auth completely optional in my extension unless you want to download IPAs — but haven't been able to reliably test the 2FA flow. Open to guidance!

As for the `Pop to Root Search - Immediately` setting... that does seem like a showstopper, unless I'm able to use long-running processes, as in the [pm2 extension](https://raycast.com/litomore/pm2)?

pernielsentikaer commented:
Can it be that @LitoMore has any ideas? Right now, it pops up the "keychain password" dialog, which closes Raycast.

LitoMore commented:
Can we save the auth step to cache? We can read the current auth step from cache when launching the extension. If there is an unfinished auth step, then we redirect to the authorization view.

What do you think?

pernielsentikaer commented:
So we're only using the 2fa/login when you run the action `Download` or do I miss anything @chrismessina 😊

chrismessina commented:
> So we're only using the 2fa/login when you run the action `Download` or do I miss anything @chrismessina 😊

Yes, the first time you use the Download command, and technically only during the Apple sign in flow for users with 2FA turned on. 

You only need to sign in to this extension _if_ you Download IPA files (which is why I built this extension in the first place, though few people may use this feature!). 

So most people may never trigger the auth flow. You can search, look up information, and download screenshots all without authenticating.

chrismessina commented:
> Just a suggestion fellas, but if the auth flow/2FA is still causing issues, I wonder if a good compromise might be removing the authentication/downloading, and instead use the app identifier to open a URL to the app instead?

Afraid not. That would literally (for me) defeat the purpose of the extension. I built this extension because I used to have to download IPAs manually; this helps me automate that process.

It'd be like asking me to remove the download capabilities of a Download Manager.

I'm proceeding with the expectation that 1) _someone_ is capable of writing code for this extension that could handle the Apple auth flows and 2) that I should bother getting this into the Raycast Store. 

If either of those are unreasonable expectations (i.e. I may _not_ be capable of writing the necessary code, even though I've really tried), let me know and I can just publish the extension code to my own Github profile and not pursue publishing to the Raycast store.

pernielsentikaer commented:
Maybe it would be a good idea to get the first iteration in the store without it, and then we can figure out a way to get it back 😊

chrismessina commented:
> Maybe it would be a good idea to get the first iteration in the store without it, and then we can figure out a way to get it back 😊

Man, this is... where the frustration builds. I'm chill, but let me explain:

1. I had a version of this extension ready to go some time back, but since it didn't handle Apple 2fa auth within the extension (you had to go to the CLI), @andreaselia asked me to implement support.
2. So, I took a month or so to add it, but it's hard to test but seems to work fine for me, and I have no useful tools or ways to test why you guys are having issues (if you are, I don't know if you're still experiencing blockers). 
3. For me to remove the download functionality would make the extension not valuable to me. Maybe it's useful for others to just be able to search for iOS apps, open them in the browser, and download screenshots, but the effort it would take to remove all that functionality is considerable, and from my lens, a big regression.

As a compromise, let's say I left in the downloading codepath but just removed the related commands:

1. Would that be acceptable, or would I need to fully remove the download functionality?
2. What guarantee would I have that the download capabilities would be restored in short order, after the first release? I'm not interested in releasing a crippled extension without some confidence that its original functionality will be restored.

@pernielsentikaer since it sounds like you and I both have a desire to get a first iteration into the store, what game plan do you suggest to address my concerns? I feel like I'm getting mixed messages from the Raycast side of the house.
